### DISCUSSION TITLE
Nextjs 14 and Chatgpt | Client side vs Server side

### DISCUSSION THREAD

**Discussion by aleksanderlukianczuk03** [vercel/next.js#62330] [Help]
### Summary

Hello, initially, I want to set up simple translation with gpt 3.5 turbo in my nextjs 14 app. When I use NEXT_PUBLIC_OPENAI_API_KEY in app/page.tsx (client side) and .env.local NEXT_PUBLIC_OPENAI_API_KEY=secret key etc. it works good but when i try to move my openai api call on a server side (for security reasons) I get page.tsx:11 POST http://localhost:3000/api/translation 404 (Not Found). I don't know exactly why as everything seem so be well done. 

Thank you very much for your help! 

Below code:  

1.
I changed `.env.local OPENAI_API_KEY=secret key`

2.
```
app/api/translation.ts
import type { NextApiRequest, NextApiResponse } from 'next';

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  if (req.method === 'POST') {
    const { input } = req.body;
    const response = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${process.env.OPENAI_API_KEY}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model: "gpt-3.5-turbo",

        messages: [
          {
            role: "system",
            content: "You are a helpful assistant."
          },
          {
            role: "user",
            content: `Translate these English words into French: ${input}`
          }
        ],
      }),
    });

    if (response.ok) {
      const data = await response.json();
      res.status(200).json({ translation: data.choices[0].message.content.trim() });
    } else {
      res.status(500).json({ error: 'Failed to fetch translation' });
    }
  } else {
    res.setHeader('Allow', ['POST']);
    res.status(405).end(`Method ${req.method} Not Allowed`);
  }
}
```
3.

app/page.tsx
```
"use client";
import Navbar from './components/navbar'
import { useState } from 'react';

const HomePage = () => {
  const [input, setInput] = useState('');
  const [translation, setTranslation] = useState('');

  const handleTranslate = async () => {
    const response = await fetch('/api/translation', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        input: input,
      }),
    });

    if (response.ok) {
      const data = await response.json();
      setTranslation(data.translation);
    } else {
      setTranslation('Error translating text. Please try again.');
    }
  };

  return (
    <div>
      <Navbar /> {/* Include the Navbar component */}
      <h1>Welcome to Language Learner!</h1>
      <h2>Translate and Learn in a Fun Way</h2>
      <textarea
        value={input}
        onChange={(e) => setInput(e.target.value)}
        placeholder="Enter words here..."
      ></textarea>
      <button onClick={handleTranslate}>Translate</button>
      <div>
        <h3>Translation:</h3>
        <p>{translation}</p>
      </div>
    </div>
  );
};

export default HomePage;
```
![Zrzut ekranu (392)](https://github.com/vercel/next.js/assets/158765636/6ef1efc5-677f-436d-99ea-6bc768385c4d)

**Comment by gustavostz** [comment]
Try to put into "pages/api/translation.ts" instead of "app/api/translation.ts"

Not sure, but I think that could be

**Reply by aleksanderlukianczuk03** [reply]
I think this doesn't work, I updated my directory as well as add POST instead of handler but still get 404 not found
![Zrzut ekranu (394)](https://github.com/vercel/next.js/assets/158765636/57a5c3e8-723e-4a53-b6ae-cb0b5fc4206f)
![Zrzut ekranu (393)](https://github.com/vercel/next.js/assets/158765636/a39501e0-12c3-4adb-991c-255f5ee8c781)
where do I make a mistake? Can you look at this code? Thank you very much

**Reply by aleksanderlukianczuk03** [reply]
thank you I used chat and completions hook: https://sdk.vercel.ai/docs/guides/frameworks/nextjs-app

**Reply by gustavostz** [reply]
Glad that you found the solution! :)

**Comment by icyJoseph** [comment]
Hi

The problem is that, in App Router, you needed to do `/api/translations/route.ts` and the contents of your file were using the Pages Router API, you had to rewrite it to use Router Handlers.

It'd have been just as simple as:

```js
// app/api/translations/route.ts

export const POST = (request) => { /* Work on the request input */ }
```
